### PR TITLE
Consider bounds when using the integral and pdf

### PR DESCRIPTION
**Pull request by chrisburr**
Corrects the usage of bounds to ensure the integral is correct and the pdf is zeroed. See [this notebook](http://nbviewer.jupyter.org/gist/chrisburr/f0590f761e9e237616c2) for and example

Fixes #36.

### CODE REVIEW

File: tensorprob/model.py [hunk 1]
```diff
@@ -67,6 +70,23 @@ def __exit__(self, e_type, e, tb):
         self.graph_ctx.__exit__(e_type, e, tb)
         self.graph_ctx = None
 
+        # Normalise all log probabilities contained in _description
+        with self._model_graph.as_default():
+            for var, (logp, integral, bounds) in self._description.items():
```
Comment (chrisburr) [inline tensorprob/model.py]: @ibab Tensor Flow fails to calculate the gradient when the log has been normalised with bounds at infinity. Replacing infinities with +/-10^250  seems to fix the problem. Can you think of any better solutions?

**Comment by ibab** [comment]
I think we can't  expect things to work in general when we supply `-inf` or `+inf` in a bound.
(Even if our current integrals work fine with it).
There's probably no way around adding special logic to deal with infinite bounds.

**Comment by chrisburr** [comment]
I suspected you'd say that. In that case this is ready to be reviewed/merged.

**Comment by ibab** [comment]
I'm not sure if replacing `+inf` and `-inf` with large positive/negative values is a good solution.
Ideally, we would just leave out those bounds.
For example, when we need to integrate from `-inf` to `0`, using the CDF we can just drop the term that handles the lower bound.

**Comment by chrisburr** [comment]
In that case I think it's better for distributions to define the `Distribution.cdf` rather than the `Distribution.integral` else we need to copy the code to check for infinities into every distribution?

**Comment by ibab** [comment]
That's right.
Maybe the `Distribution.integral` should assume that we want an unconstrained integral when we pass `None` as one of the integration limits?
As in `integral(None, 0)` integrating from `-inf` to `0`.
But I think we can postpone these things to a later change.
I can merge this now if you're okay with it.

**Comment by ibab** [comment]
Or maybe `-inf` and `+inf` should be used instead of `None` and it's simply the job of the `Distribution.integral` to provide something that can be differentiated?

**Comment by chrisburr** [comment]
I'm happy for you to merge it. :+1: 

I prefer `-inf` and `+inf` and requiring `Distribution.integral` to provide something that can be differentiated is probably the best way forward. I'll try testing a few ideas in a new branch.

**Pull request merged**